ansi-c: explicit casts from structs to structs

kroening opened the pull request:
The C frontend should allow casting a struct to itself using a typecast.

Reported as issue #5153.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [X] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [X] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- n/a My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: eec84a6dc62f0e932f1c8b01f9f46cf236bd0495).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/130998753

peterschrammel commented on regression/ansi-c/typecasts/typecasts1.desc:
```diff
@@ -0,0 +1,9 @@
+--
+^CONVERSION ERROR$
+--
+Reported as isse #5153
```
```suggestion
Reported as issue #5153
```

kroening replied:
fixed

peterschrammel reviewed (commented):
Why is there java-models-library bump in the second commit?

kroening commented:
> Why is there java-models-library bump in the second commit?

Fixed

kroening committed:
ansi-c: do not use expansion of tags when checking typecasts

The C frontend should allow casting a struct to itself using a typecast.
Reported as issue #5153.

This commit adds a test and fixes the problem.

smowton reviewed (commented):
So the test casts from a struct to the same struct, but the code change solely relates to union types? How's that work?

kroening commented:
The piece of code changed handles both unions and structs. The important bit is not to expand the tag to the type, which implies the need for expanding it later when looking for the first component of the union.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 5fa41b9501f425223123ea1664777de7d7cbb876).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/131165614
